cv_bridge/src/cv_mat_sensor_msgs_image_type_adapter.cpp

```diff
@@ -76,46 +86,36 @@ ROSCvMatContainer::ROSCvMatContainer(
       unique_sensor_msgs_image.get(),
       "unique_sensor_msgs_image cannot be nullptr"
 ).pointer->header),
-  frame_(
-    unique_sensor_msgs_image->height,
-    unique_sensor_msgs_image->width,
-    encoding2mat_type(unique_sensor_msgs_image->encoding),
-    unique_sensor_msgs_image->data.data(),
-    unique_sensor_msgs_image->step),
+  frame_(matFromImage(*unique_sensor_msgs_image)),
   storage_(std::move(unique_sensor_msgs_image))
 {}
 
 ROSCvMatContainer::ROSCvMatContainer(
   std::shared_ptr<sensor_msgs::msg::Image> shared_sensor_msgs_image)
 : header_(shared_sensor_msgs_image->header),
-  frame_(
-    shared_sensor_msgs_image->height,
-    shared_sensor_msgs_image->width,
-    encoding2mat_type(shared_sensor_msgs_image->encoding),
-    shared_sensor_msgs_image->data.data(),
-    shared_sensor_msgs_image->step),
+  frame_(matFromImage(*shared_sensor_msgs_image)),
   storage_(shared_sensor_msgs_image)
 {}
 
 ROSCvMatContainer::ROSCvMatContainer(
-  const cv::Mat & mat_frame,
+  cv::Mat mat_frame,
   const std_msgs::msg::Header & header,
   bool is_bigendian,
   std::optional<std::string> encoding_override)
 : header_(header),
-  frame_(mat_frame),
+  frame_(std::move(mat_frame)),
   storage_(nullptr),
   is_bigendian_(is_bigendian),
   encoding_override_(encoding_override)
 {}
 
 ROSCvMatContainer::ROSCvMatContainer(
-  cv::Mat && mat_frame,
+  cv::cuda::GpuMat mat_frame,
   const std_msgs::msg::Header & header,
   bool is_bigendian,
   std::optional<std::string> encoding_override)
```

Comment:
Probably should not call it encoding_override but encoding. As it specifies the stored encoding or guess from cv type if empty.

Reply:
I think the target usage here is different. If set, it will ignore the encoding of ```cv::Mat``` and set that as the encoding of the message type. Could be useful since from what I can tell, when they serialize ```cv::Mat``` they do not mix the channels so _"fix"_ the issues with ```RGB``` vs ```BGR``` they leave this field for the user to decide which encoding to publish. 

It is, however, more efficient to serialize it like this. Though I'm not totally sold on it vs, let's say, having a custom ```cv_RGB``` encoding so it's properly signaled downstream.